lib/src/local_working_copy.rs

```diff
@@ -671,6 +684,56 @@ impl TreeState {
         Ok(())
     }
 
+    fn save_only_clock(&mut self) -> Result<(), TreeStateError> {
+        let mut temp_file = NamedTempFile::new_in(&self.state_path).unwrap();
+        let target_path = self.state_path.join("tree_state");
+
+        // If the same non-repeated field appears multiple times, only the last will be used:
+        // https://protobuf.dev/programming-guides/encoding/#last-one-wins
+        // Reuse the existing serialisation of the proto in the current state to avoid
+        // serialising non-changed fields.
+        let bytes_written = std::fs::copy(&target_path, temp_file.path()).map_err(|err| {
```

Comment:
> even if it could, it's safe to rewind the watchman clock.

Yes, that is correct. From @arxanas:

> [It's not ever incorrect to ignore the clock update, but then you have to redo all the same incremental processing work on the next query. (If there is no additional work, then the inefficiency would be minimal.)](https://discord.com/channels/968932220549103686/969291218347524238/1245965414857707540)

We should still save the watchman clock due to this "incremental processing work" - for example, consider a large build directory that is jj-ignored but not .watchmanconfig ignored.

***

> it might be simpler to extract `watchman_clock` to separate proto file if the second bullet point holds.

This matches @arxanas' idea of splitting the states:

> [Another alternative is to split the O(repo) working copy state from the O(1) state and invalidate/update them separately. Then you could update the clock without having to rewrite the whole file state map.](https://discord.com/channels/968932220549103686/969291218347524238/1245968572803579905)

I think this is quite reasonable, but I was a bit concerned about the atomicity of having two working copy protos. However, this might not be a problem?

> [I guess it's possible that you'd want to write all that information atomically somehow? I don't see when it would matter, though, at least for the file states map, which IIRC is just an optimization and not incorrect to fail at updating?](https://discord.com/channels/968932220549103686/969291218347524238/1245969222522376332)  

What do you think about splitting the working state in general? I think that going forward, this might be easier to maintain than special-casing the clock in this PR.

Reply:
I don't think atomic update in general is required here (because we'll take a lock and reload on mutation, and actual working-copy files can't be updated atomically anyway), but I'm not pretty sure. @martinvonz, any concerns?

From @arxanas' comment, I think it's safe to split `watchman_clock` to a separate file. The combination of `(new_tree_state, old_watchman_clock)` won't cause consistency problem.

Reply:
I agree - as long as we can't end up writing `(old_tree_state, new_watchman_clock)`, there shouldn't be any problem.

Reply:
I was slightly worried about the following situation:

- Time 0: `foo.txt` has contents `1`. (`old_watchman_clock`)
- Time 1: `foo.txt` has contents `2`. (`new_tree_state`)
- Time 2: `foo.txt` has contents `1`.

At time 2, querying watchman with the time 0 clock successfully shows `foo.txt` being updated, even though its contents technically have not been updated.

I'll abandon this PR and will work on splitting out the proto. Just a few questions before I get started:

- Splitting - what part of `TreeState` should we "split out"? I think there's two reasonable options:
  - Move only `watchman_clock` into a new message, as discussed. I think this is most reasonable as an incremental update to `TreeState`.
  - Move only `file_states` into a new message. This notably allows us to update `tree_ids` and `sparse_patterns` without writing `file_states`, but I think this is probably not needed.
- Naming - what should we name the new message / the on-disk file? Thankfully, protobuf makes it easy to rename messages, so we don't need to think about this _too_ much.
  - If moving `watchman_clock`: Should we keep it tied with the idea of a file system watcher, or keep it more general? I think `TreeStateSecondary` is reasonable - it emphasises the fact that a "primary" update can be done without updating the "secondary".
  - If moving `file_states`: I think it's reasonable for it to be called `FileStates`.

Reply:
> At time 2, querying watchman with the time 0 clock successfully shows foo.txt being updated, even though its contents technically have not been updated.

That should just be a performance problem, not a correctness problem. It would mean that we read the file contents an extra time before we determine that it hasn't changed.

> * Move only `file_states` into a new message. This notably allows us to update `tree_ids` and `sparse_patterns` without writing `file_states`, but I think this is probably not needed.

More importantly, those needs to be written atomically.

> Naming - what should we name the new message / the on-disk file?

I would go with something more specific since we don't yet know what else might go into that file.

Reply:
I was dealing with the same problem at work recently, specifically concerned with the case that the user creates a new file at time 0, we process it at time 1 and then abort, and then the user deletes the file at time 2, so that the working copy state compared to the last saved clock looks identical to us, but we created extra state somewhere that now needs to be cleaned up. But Watchman will still report the file change so that we can notice and delete the extra state.